File: src/test/java/ar/edu/unlam/tallerweb1/infrastructure/UserJdbcRepositoryTest.java
package ar.edu.unlam.tallerweb1.infrastructure;

import ar.edu.unlam.tallerweb1.domain.UserRepository;
import ar.edu.unlam.tallerweb1.domain.model.FriendsGroup;
import ar.edu.unlam.tallerweb1.domain.model.User;
import org.junit.jupiter.api.Test;

import java.sql.SQLException;

import static ar.edu.unlam.tallerweb1.infrastructure.utils.TestsPersistenceHelper.createGroup;
import static ar.edu.unlam.tallerweb1.infrastructure.utils.TestsPersistenceHelper.getFriendsGroupId;
import static org.assertj.core.api.Assertions.assertThat;

public class UserJdbcRepositoryTest extends JdbcRepositoryTest {

    private UserRepository userRepository = new UserJdbcRepostory(dataSource);
    private final String USER = "user_18989";
    private final String GROUP = "group_sfaf1";
    @Test
    public void createANewUser() {
        whenCreateAnUser(USER);
        thenCanFindIt(USER);
    }

    @Test
    public void createANewUserWithFriendGroup() throws SQLException {
        createGroup(connection, GROUP);
        whenCreateAnUser(USER, GROUP);
        thenCanFindItAndHasAFriendsGroup(USER);
    }

    private void whenCreateAnUser(String user, String group) throws SQLException {
        final Long friendsGroupId = getFriendsGroupId(connection, group);
        User userToBeCreated = new User(user);
        userToBeCreated.setFriendsGroup( new FriendsGroup(new Long(friendsGroupId), group));
        userRepository.saveWithFriendGroup(userToBeCreated);
    }

    @Test
    public void searchAnUnexistingUser(){
        whenNotExistsTheUser(USER);
        thenCantFindIt(USER);
    }

    private void whenNotExistsTheUser(String user) {
    }

    private void whenCreateAnUser(String user) {
        userRepository.save(new User(user));
    }

    private void thenCantFindIt(String user) {
        assertThat(userRepository.findByName(user)).isNull();
    }

    private void thenCanFindItAndHasAFriendsGroup(String user) {
        final User found = userRepository.findByName(user);
        assertThat(found).isNotNull();
        assertThat(found.getFriendsGroup()).isNotNull();
    }

    private void thenCanFindIt(String user) {
        assertThat(userRepository.findByName(user)).isNotNull();
    }
}

File: src/test/java/ar/edu/unlam/tallerweb1/infrastructure/UserJdbcRepositoryTest.kt
package ar.edu.unlam.tallerweb1.infrastructure

import ar.edu.unlam.tallerweb1.domain.UserRepository
import ar.edu.unlam.tallerweb1.domain.model.FriendsGroup
import ar.edu.unlam.tallerweb1.domain.model.User
import ar.edu.unlam.tallerweb1.infrastructure.utils.TestsPersistenceHelper.createGroup
import ar.edu.unlam.tallerweb1.infrastructure.utils.TestsPersistenceHelper.getFriendsGroupId
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.sql.SQLException

class UserJdbcRepositoryTest : JdbcRepositoryTest() {
    private val userRepository: UserRepository = UserJdbcRepostory(dataSource)
    private val USER = "user_18989"
    private val GROUP = "group_sfaf1"
    @Test
    fun createANewUser() {
        whenCreateAnUser(USER)
        thenCanFindIt(USER)
    }

    @Test
    @Throws(SQLException::class)
    fun createANewUserWithFriendGroup() {
        createGroup(connection, GROUP)
        whenCreateAnUser(USER, GROUP)
        thenCanFindItAndHasAFriendsGroup(USER)
    }

    @Throws(SQLException::class)
    private fun whenCreateAnUser(user: String, group: String) {
        val friendsGroupId = getFriendsGroupId(connection, group)
        val userToBeCreated = User(user)
        userToBeCreated.friendsGroup = FriendsGroup(friendsGroupId, group)
        userRepository.saveWithFriendGroup(userToBeCreated)
    }

    @Test
    fun searchAnUnexistingUser() {
        whenNotExistsTheUser(USER)
        thenCantFindIt(USER)
    }

    private fun whenNotExistsTheUser(user: String) {}

    private fun whenCreateAnUser(user: String) = userRepository.save(User(user))

    private fun thenCantFindIt(user: String) = assertThat(userRepository.findByName(user)).isNull()

    private fun thenCanFindItAndHasAFriendsGroup(user: String) {
        val found = userRepository.findByName(user)
        assertThat(found).isNotNull()
        assertThat(found!!.friendsGroup).isNotNull()
    }

    private fun thenCanFindIt(user: String) = assertThat(userRepository.findByName(user)).isNotNull()
}
